### PR TITLE
[WIP] Brigadier Command Api v3

### PR DESCRIPTION
**Pull request by hYdos**
Can't use V2 just in case someone actually used v2 in their mod in the short time it was up.

### CODE REVIEW

**Comment by PseudoDistant** [comment]
I rescind my f den.

File: build.gradle [hunk 1]
```diff
@@ -84,14 +84,14 @@ allprojects {
 	}
 
 	dependencies {
+		implementation("com.mojang:brigadier:1.0.17")
```
Comment (BoogieMonster1O1) [inline build.gradle:87]: this doesn't include brigadier

File: modules/base/src/main/java/net/legacyfabric/fabric/impl/base/mixin/RealmsBridgeMixin.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+@Mixin(RealmsBridge.class)
+public class RealmsBridgeMixin {
```
Comment (BoogieMonster1O1) [inline modules/base/src/main/java/net/legacyfabric/fabric/impl/base/mixin/RealmsBridgeMixin.java:16]: 👎

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		return name;
+	}
+
+	@Override
+	public List<String> getAliases() {
+		return Collections.emptyList();
+	}
+
+	@Override
+	public void execute(CommandSource source, String[] args) {
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java:38]: why's this empty

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+	@Override
+	public List<String> getAutoCompleteHints(CommandSource source, String[] args, BlockPos pos) {
+		//TODO: autocomplete suggestions
+		return Collections.emptyList();
+	}
+
+	@Override
+	public boolean isUsernameAtIndex(String[] args, int index) {
+		return false;
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java:54]: fix this

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	@Override
+	public int compareTo(@NotNull Command o) {
+		return 0;
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java:59]: this as well

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import com.mojang.brigadier.tree.CommandNode;
+import net.minecraft.command.Command;
+import net.minecraft.command.CommandSource;
+import net.minecraft.util.math.BlockPos;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collections;
+import java.util.List;
+
+public class BrigadierCommand implements Command {
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/BrigadierCommand.java:12]: don't implement `Command`
extend `AbstractCommand`

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/CommandRegisterCallback.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	});
+
+	@FunctionalInterface
+	public interface CommonCommandRegister {
+		void onCommandRegistration(CommandDispatcher<CommandSource> dispatcher, Profiler profiler);
+	}
+
+	@FunctionalInterface
+	public interface ClientCommandRegister {
+		void onCommandRegistration(CommandDispatcher<CommandSource> dispatcher, MinecraftClient client, Profiler profiler);
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/CommandRegisterCallback.java:89]: unsafe usage of `MinecraftClient` on the server thread

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/CommandRegisterCallback.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+			}
+		}
+	});
+
+	/**
+	 * Called when it is time to register Common sided commands.
+	 *
+	 * <p>When this event is called, all Vanilla commands have already been registered
+	 */
+	public static final Event<ClientCommandRegister> CLIENT = EventFactory.createArrayBacked(ClientCommandRegister.class, callbacks -> (dispatcher, client, profiler) -> {
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/CommandRegisterCallback.java:41]: `CLIENT` is misleading
should be renamed to integrated

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/CommandRegisterCallback.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			}
+		}
+	});
+
+	/**
+	 * Called when it is time to register Common sided commands.
+	 *
+	 * <p>When this event is called, all Vanilla commands have already been registered
+	 */
+	public static final Event<ServerCommandRegister> SERVER = EventFactory.createArrayBacked(ServerCommandRegister.class, callbacks -> (dispatcher, server, profiler) -> {
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/CommandRegisterCallback.java:64]: ditto
dedicated

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/mixin/CommandManagerMixin.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import java.util.List;
+
+@Mixin(CommandManager.class)
+public abstract class CommandManagerMixin extends CommandRegistry {
+
+	@Inject(method = "<init>", at = @At("TAIL"))
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/mixin/CommandManagerMixin.java:23]: you don't need a mixin for this

File: modules/commands/src/main/java/net/legacyfabric/fabric/commands/mixin/CommandManagerMixin.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	/**
+	 * Returns all commands registered and accessible to a source. Overwritten to include Brigadier commands.
+	 *
+	 * @param source the source of the command
+	 * @return a list of available commands to the provided source.
+	 */
+	@Override
+	public List<Command> method_5983(CommandSource source) {
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/java/net/legacyfabric/fabric/commands/mixin/CommandManagerMixin.java:42]: 👎 
there are better ways to do this
I did it with sponge cmd api without a massive hack

File: modules/commands/src/main/resources/fabric.mod.json [hunk 12]
```diff
@@ -0,0 +1,24 @@
+{
+  "schemaVersion": 1,
+  "id": "legacy-fabric-commands-v3",
+  "name": "Legacy Fabric Commands (v3)",
+  "version": "${version}",
+  "environment": "*",
+  "license": "Apache-2.0",
+  "icon": "assets/legacy-fabric/icon.png",
+  "contact": {
+    "homepage": "https://legacy-fabric.github.io/",
```
Comment (BoogieMonster1O1) [inline modules/commands/src/main/resources/fabric.mod.json:10]: wrong homepage

File: settings.gradle [hunk 13]
```diff
@@ -25,12 +25,13 @@ def addModule(String dir, String subdir, String name) {
 
 addModule("base", "legacy-fabric-api-base")
 addModule("resources", "legacy-fabric-resource-loader-v1")
-addModule("client", "rendering", "legacy-fabric-rendering-api-v1")
-addModule("client", "keybindings", "legacy-fabric-keybinding-api-v1")
 addModule("networking", "legacy-fabric-networking-api-v1")
+addModule("commands", "legacy-fabric-commands-v3")
```
Comment (BoogieMonster1O1) [inline settings.gradle:29]: brigadier-commands

**Review by BoogieMonster1O1 — CHANGES_REQUESTED** [review]
rename the module to mention brigadier
rename the packages

**Comment by ramidzkh** [comment]
So what happened to this https://github.com/Legacy-Fabric/fabric/pull/99#issuecomment-980840833

**Comment by hYdos** [comment]
I made an compromise. This was the best option as it allows for portability and ease of use.

**Comment by ramidzkh** [comment]
So when are you pushing the changes you resolved the reviews for

**Comment by arthurbambou** [comment]
stale

**Pull request closed without merging**